src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs

```diff
@@ -306,20 +380,98 @@ public async Task UpdateEditorAsync(ContentItem contentItem, UpdateEditorContext
             foreach (var typePartDefinition in contentTypeDefinition.Parts)
             {
                 var partTypeName = typePartDefinition.PartDefinition.Name;
+                var partName = typePartDefinition.Name;
+                var contentType = typePartDefinition.ContentTypeDefinition.Name;
                 var activator = _contentPartFactory.GetTypeActivator(partTypeName);
-                var part = (ContentPart)contentItem.Get(activator.Type, typePartDefinition.Name) ?? activator.CreateInstance();
+                var part = (ContentPart)contentItem.Get(activator.Type, partName) ?? activator.CreateInstance();
+                var isNamedPart = typePartDefinition.PartDefinition.IsReusable() && partName != partTypeName;
+                var partPosition = typePartDefinition.GetSettings<ContentTypePartSettings>().Position ?? "before";
                 part.ContentItem = contentItem;
 
                 // Create a custom shape to render all the part shapes into it
-                var typePartShape = await context.ShapeFactory.CreateAsync("ContentPart_Edit");
+                var shapeType = "ContentPart_Edit";
+                var typePartShapeResult = new ShapeResult(shapeType, ctx => ctx.ShapeFactory.CreateAsync(shapeType));
+
+                typePartShapeResult.Differentiator($"{contentType}-{partName}");
+                typePartShapeResult.Name(partName);
+                typePartShapeResult.Location($"Parts:{partPosition}");
+
+                typePartShapeResult.Displaying(ctx =>
+                {
+                    // ContentPart_Edit__[PartType]
+                    // eg ContentPart.ServicePart.Edit
+                    ctx.Shape.Metadata.Alternates.Add($"{shapeType}__{partTypeName}");
+
+                    // ContentPart_Edit_[ContentType]_[PartType]
+                    // e.g. ContentPart-LandingPage-ServicePart.Edit
+                    ctx.Shape.Metadata.Alternates.Add($"{shapeType}__{contentType}__{partTypeName}");
+
+                    if (isNamedPart)
+                    {
+                        // ContentPart_Edit__[ContentType]__[PartName]
+                        // e.g. ContentPart-LandingPage-BillingService.Edit ContentPart-LandingPage-HelplineService.Edit
+                        ctx.Shape.Metadata.Alternates.Add($"{shapeType}__{contentType}__{partName}");
+                    }
+                });
+
+                await typePartShapeResult.ApplyAsync(context);
+                var typePartShape = typePartShapeResult.Shape;
+
+                if (typePartShape == null)
+                {
+                    // Part is explicitly hidden in placement then stop rendering it
+                    continue;
+                }
+
                 typePartShape.Properties["ContentPart"] = part;
                 typePartShape.Properties["ContentTypePartDefinition"] = typePartDefinition;
-                var partPosition = typePartDefinition.GetSettings<ContentTypePartSettings>().Position ?? "before";
```

Comment:
Looks like we just need to have 

```
                    nameResult.Differentiator($"{contentType}-{partName}");
                    nameResult.Name(partName);
                    nameResult.Location($"Parts.{partName}.Header");
```

added. Plus the alternates.

I am not sure about the necessity of the new wrapping shapes and zone.

Reply:
yes this line is moved above `ApplyAsyc`